### PR TITLE
Encode delta index epoch

### PR DESCRIPTION
**Pull request by breezewish**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/6827

Problem Summary:

### What is changed and how it works?

Introduce `delta_index_epoch`, which indicates that this delta index instance cannot be used for incremental update. Currently only `flush()` will cause `delta_index_epoch++`.

TODO: May be a better way is that we could detect missing updates and drop the index internally?

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/Delta/DeltaValueSpace.cpp [hunk 1]
```diff
@@ -343,8 +343,14 @@ bool DeltaValueSpace::flush(DMContext & context)
 
         /// Update delta tree
         if (new_delta_index)
+        {
             delta_index = new_delta_index;
 
+            // Indicate that the index with old epoch should not be used anymore.
+            // This is useful in disaggregated mode which will invalidate the delta index cache in RN.
+            delta_index_epoch += 1;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Delta/DeltaValueSpace.cpp:351]: Seems there is no synchronization between `DeltaValueSpace::createSnapshot` (for_update=false) and `DeltaValueSpace::flush`. Is there a chance that we still get an incorrect delta-index?

Comment (breezewish) [reply dbms/src/Storages/DeltaMerge/Delta/DeltaValueSpace.cpp:351]: Good catch. We should ensure the delta_index and delta_index_epoch are either updated, or either not updated.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**